File: namastack-outbox-examples/namastack-outbox-example-testcontainers-stable/src/main/java/io/namastack/demo/stabletest/KafkaOutboxRoutingConfiguration.java
package io.namastack.demo.stabletest;

import io.namastack.demo.stabletest.customer.CustomerRegisteredEvent;
import io.namastack.outbox.kafka.KafkaOutboxRouting;
import io.namastack.outbox.routing.selector.OutboxPayloadSelector;
import java.util.Map;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Configuration for Kafka outbox routing using the Java builder API.
 * <p>
 * This demonstrates how to configure routing rules in pure Java: - Route CustomerRegisteredEvent to
 * a specific topic with custom headers - Use defaults for all other events
 */
@Configuration
public class KafkaOutboxRoutingConfiguration {

  @Bean
  public KafkaOutboxRouting kafkaOutboxRouting(  @Value("${demo.topic}") String demoTopic) {
    return KafkaOutboxRouting.builder()
        .route(OutboxPayloadSelector.type(CustomerRegisteredEvent.class), route -> {
          route.target(demoTopic);
          route.key((payload, metadata) -> metadata.getKey());
          route.headers((payload, metadata) -> Map.of(
              "CustomerMail", ((CustomerRegisteredEvent) payload).getEmail()));
        })
        .defaults(route -> {
          route.target("default-topic");
          route.key((payload, metadata) -> metadata.getKey());
          route.headers((payload, metadata) -> Map.of(
              "eventType", payload.getClass().getSimpleName()
          ));
        })
        .build();
  }
}

File: namastack-outbox-examples/namastack-outbox-example-testcontainers-flaky/src/main/kotlin/io/namastack/demo/flakytest/KafkaOutboxRoutingConfiguration.kt
package io.namastack.demo.flakytest

import io.namastack.demo.flakytest.customer.CustomerRegisteredEvent
import io.namastack.outbox.kafka.KafkaOutboxRouting
import io.namastack.outbox.kafka.kafkaOutboxRouting
import io.namastack.outbox.routing.selector.OutboxPayloadSelector.Companion.type
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class KafkaOutboxRoutingConfiguration {

    @Bean
    fun kafkaRouting(@Value($$"${demo.topic}") demoTopic: String): KafkaOutboxRouting =
        kafkaOutboxRouting {
            route(type(CustomerRegisteredEvent::class.java)) {
                target(demoTopic)
                key { _, metadata -> metadata.key }
                headers { payload, _ ->
                    mapOf(
                        "CustomerMail" to (payload as CustomerRegisteredEvent).email,
                    )
                }
            }
            defaults {
                target("default-topic")
                key { _, metadata -> metadata.key }
                headers { payload, _ -> mapOf("eventType" to payload.javaClass.simpleName) }
            }
        }
}
